### PR TITLE
[Deprecated][Dynamo] Enhanced onnxrt backend

### PR DESCRIPTION
**Pull request by fffrog**
1. onnxrt backend raise error when passing options parameter
2. onnxrt support more detailed EP configure capability by provider parameter

```python
import torch
import torch._dynamo

providers = [
    (
        "CUDAExecutionProvider",
        {
            "device_id": 0,
            "arena_extend_strategy": "kNextPowerOfTwo",
            "gpu_mem_limit": 2 * 1024 * 1024 * 1024,
            "do_copy_in_default_stream": True,
        },
    ),
    "CPUExecutionProvider",
]


@torch.compile(backend="onnxrt", options={"providers": providers})
def foo(x, y):
    return (x + y) * x


if __name__ == "__main__":
    a, b = torch.randn(10), torch.ones(10)
    print(foo(a, b))
```

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

**Comment by FFFrog** [comment]
Sorry to bother @Chillee, could you please to take a look at this pr? Is there anything else needed to make it ok?

**Comment by FFFrog** [comment]
@wschin, @thiagocrepaldi, can you guys please to take a look at this pr? Any changes needed to do?

File: test/dynamo/test_backends.py [hunk 1]
```diff
@@ -142,6 +142,24 @@ def test_nvprims_aten(self):
     def test_onnxrt(self):
         self._check_backend_works("onnxrt")
 
+    @requires_cuda()
+    @unittest.skipIf(not has_onnxruntime(), "requires onnxruntime")
+    def test_onnxrt_with_options(self):
```
Comment (thiagocrepaldi) [inline test/dynamo/test_backends.py:147]: Maybe add a test with `options=None` to guard against future regressions?

File: torch/_dynamo/backends/onnxrt.py [hunk 2]
```diff
@@ -78,10 +85,17 @@ def onnxrt(gm, example_inputs, *, filename=None, provider=None):
...
-        provider = default_provider(device_type)
-    assert provider in onnxruntime.get_available_providers()
-    session = onnxruntime.InferenceSession(filename, providers=[provider])
+    providers = options.get("providers") if options else None
+    if providers is None:
+        providers = [default_provider(device_type)]
+
+    for provider in providers:
+        if isinstance(provider, str):
+            assert provider in onnxruntime.get_available_providers()
```
Comment (thiagocrepaldi) [inline torch/_dynamo/backends/onnxrt.py:94]:
```suggestion
            assert provider in onnxruntime.get_available_providers(), "Some meaningful error message"
```

File: torch/_dynamo/backends/onnxrt.py [hunk 3]
```diff
@@ -78,10 +85,17 @@ def onnxrt(gm, example_inputs, *, filename=None, provider=None):
...
+        elif isinstance(provider, tuple):
+            assert provider[0] in onnxruntime.get_available_providers()
```
Comment (thiagocrepaldi) [inline torch/_dynamo/backends/onnxrt.py:96]:
```suggestion
            assert provider[0] in onnxruntime.get_available_providers(), "Some meaningful error message"
```

File: torch/__init__.py [hunk 4]
```diff
@@ -1649,6 +1649,7 @@ def compile(model: Optional[Callable] = None, *,
         - `trace.enabled` which is the most useful debugging flag to turn on
         - `trace.graph_diagram` which will show you a picture of your graph after fusion
         - For inductor you can see the full list of configs that it supports by calling `torch._inductor.list_options()`
+        - For onnxrt `providers` which will be useful to configure detailed EP information
```
Comment (thiagocrepaldi) [inline torch/__init__.py:1652]:
```suggestion
        - For onnxrt `providers` which will be useful to configure detailed EP information. This is a list of `ORTExecutionProviderClass`
```
A link to the EP class/docstring would help users learn how to use this

**Review by thiagocrepaldi — COMMENTED** [review]
LGTM. I just nit picked some UX stuff

**Comment by thiagocrepaldi** [comment]
https://github.com/pytorch/pytorch/pull/106929 might superseed this PR because a new onnxrt backend might be introduced. Check with @wschin

**Comment by wschin** [comment]
> #106929 might superseed this PR because a new onnxrt backend might be introduced. Check with @wschin

Yes, the legacy `onnxrt` will be deprecated by a more-dynamo friendly `OrtBackend` introduced in https://github.com/pytorch/pytorch/pull/106589 (originally developed in https://github.com/microsoft/onnxruntime/pull/13259).

**Comment by thiagocrepaldi** [comment]
@FFFrog Check whether the changes you are proposing here are applicable after https://github.com/pytorch/pytorch/pull/106929 replaces the onnxrt implementation.

If it does, then you can close this one

**Comment by FFFrog** [comment]
> @FFFrog Check whether the changes you are proposing here are applicable after #106929 replaces the onnxrt implementation.
> 
> If it does, then you can close this one

Got it, Thank you for your relpy, I will close this one.

**Pull request closed without merging**